### PR TITLE
Implement FLIP #43 (fn type syntax change)

### LINKED ISSUE #2094: Change syntax for function types
### Issue To Be Solved

Currently, function types are printed with confusing syntax. For instance, the type of the function

```cadence
fun foo(n: Int, s: String): Bool
```

is rendered as

```cadence
foo: ((Int, String): Bool)
```

This inconsistency is more apparent when writing higher-order functions:

```cadence
fun forEach(xs: [Int], f: ((Int): Bool)): Void
```

which harms readability.

### Suggested Solution

Function types should use the `fun` keyword so their types are closer to the way they're defined. For the examples above, their types should be rendered as

```cadence
foo: fun(Int, String): Bool
forEach: fun([Int], fun(Int): Bool): Void
```

To avoid breaking existing contracts, this could be implemented as a non-breaking change by allowing the old syntax but updating the parser to also accept the `fun` keyword as a prefix. The stringification of function types should be changed to include the `fun` keyword. 

[See FLIP](https://github.com/onflow/flips/pull/43)

### PR DESCRIPTION
**Pull request by dreamsmasher**
Closes #2094, [FLIP #43](https://github.com/onflow/flips/pull/43)

## Description

Changes function types to make use of the `fun` keyword. For example:

```cadence
fun foo(x: Int8): Int16 {...}

// old syntax
let f: ((Int8): Int16) = foo

// new syntax
let g: fun(Int8): Int16 = foo
```

The old syntax is no longer supported, and pretty-printing will now make use of the new syntax. This PR also extends the parser to allow parenthesized types: for a type `T`, `(T)` will simply parse the inner type. This was previously restricted only to function types, but users can now parenthesize anything. ٩(◕‿◕｡)۶

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/ast/type.go [hunk 1]
```diff
@@ -462,19 +462,28 @@ func (t *FunctionType) String() string {
...
+
+	if t.PurityAnnotation != FunctionPurityUnspecified {
+		result = append(
+			result,
+			prettier.Text(t.PurityAnnotation.Keyword()),
+			prettier.Space,
+		)
 	}
 
+	result = append(result, functionTypeKeywordDoc)
```
Comment (turbolent) [inline runtime/ast/type.go:485]: It would be good to add a space after the keyword, like for all other keywords

**Comment by turbolent** [comment]
Please also add a label. For documentation, we probably do not want to switch to the new syntax yet once it is available. Could you please create a follow-up issue to update the documentation later?

File: runtime/parser/expression.go [hunk 2]
```diff
@@ -816,20 +816,20 @@ func defineIdentifierExpression() {
...
-						p.current = current
-					}
+				cursor := p.tokens.Cursor()
+
+				p.skipSpaceAndComments()
+
+				if p.isToken(p.current, lexer.TokenIdentifier, KeywordFun) {
+					// skip the `fun` keyword
+					p.nextSemanticToken()
+					return parseFunctionExpression(p, token, ast.FunctionPurityView)
```
Comment (dreamsmasher) [inline runtime/parser/expression.go:827]: noticed a bug when resolving the merge conflict - by only checking for a space after the `view` keyword, then expressions like

```cadence
let x = view /* UwU */ fun() {}
```

fails to parse.

File: runtime/parser/type.go [hunk 3]
```diff
@@ -148,39 +148,9 @@ func init() {
...
-					right,
-					token.StartPos,
-				), nil
-
-			default:
-				return parseNominalTypeRemainder(p, token)
-			}
-		},
-	)
+	defineOldSyntaxFunctionType()
```
[the remark this answers is not in the document]
Comment (dreamsmasher) [reply runtime/parser/type.go:152]: ah ok, like allow the new syntax in `master` then deprecate it for stable cadence? this PR does change the pretty-printing of function types too, i'm worried that might be a breaking change. thoughts?

File: runtime/parser/expression.go (hunk 2 above)
Comment (turbolent) [reply runtime/parser/expression.go:827]: Good catch! Could you please open this fix as a separate PR?

File: runtime/parser/type.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (turbolent) [reply runtime/parser/type.go:152]: The easiest is to just target Stable Cadence and remove support for the old syntax.

**Comment by turbolent** [comment]
@dreamsmasher Still missing documentation updates

**Review by turbolent — APPROVED** [review]

**Comment by turbolent** [comment]
@SupunS @dsainati1 could you please have a look so we can get this merged? Thank you

**Review by SupunS — APPROVED** [review]

**Pull request merged**